File: sdk/binding/android/q-voice-java/src/main/java/com/qdreamer/qvoice/QModule.java
package com.qdreamer.qvoice;

import com.qdreamer.qvoice.listener.OnQVoiceCallback;

import java.util.Arrays;

public class QModule {

    private static final long MODULE_VALUE = 0L;

    private OnQVoiceCallback qVoiceCallback;
    private long moduleValue = MODULE_VALUE;
    private boolean isModuleStarted = false;
    private Thread moduleThread = null;

    private native long moduleNew(long sessionValue, String param, String resPath);
    private native void moduleDel(long moduleValue);
    private native int moduleStart(long moduleValue);
    private native int moduleSet(long moduleValue, String param);
    private native int moduleStop(long moduleValue);
    private native byte[] moduleRead(long moduleValue);
    private native int moduleFeed(long moduleValue, byte[] data, int len);

    public QModule(long sessionValue, String param, String resPath, OnQVoiceCallback callback) {
        this.qVoiceCallback = callback;

        moduleValue = moduleNew(sessionValue, param, resPath);
        moduleThread = new Thread(() -> {
            try {
                while (moduleValue != MODULE_VALUE) {
                    byte[] bytes = moduleRead(moduleValue);
                    if (qVoiceCallback != null && moduleValue != MODULE_VALUE && bytes != null && bytes.length > 0) {
                        if (bytes.length > 1) {
                            qVoiceCallback.onQVoiceCallback(bytes[0], Arrays.copyOfRange(bytes, 1, bytes.length));
                        } else {
                            qVoiceCallback.onQVoiceCallback(bytes[0], null);
                        }
                    }
                    Thread.sleep(1);
                }
            } catch (Exception ignore) {
            }
        });
        moduleThread.start();
    }

    public boolean startModule() {
        synchronized (this) {
            if (moduleValue != MODULE_VALUE && !isModuleStarted) {
                moduleStart(moduleValue);
                isModuleStarted = true;
            }
            return isModuleStarted;
        }
    }

    public void feedModule(byte[] audio) {
        if (moduleValue != MODULE_VALUE && isModuleStarted) {
            if (audio == null) {
                moduleFeed(moduleValue, new byte[0], 0);
            } else {
                moduleFeed(moduleValue, audio, audio.length);
            }
        }
    }

    public boolean stopModule() {
        synchronized (this) {
            if (moduleValue != MODULE_VALUE && isModuleStarted) {
                moduleStop(moduleValue);
                isModuleStarted = false;
            }
            return !isModuleStarted;
        }
    }

    public void releaseModule() {
        stopModule();
        qVoiceCallback = null;
        long temp = moduleValue;
        moduleValue = MODULE_VALUE;
        if (moduleThread != null && moduleThread.isAlive() && !moduleThread.isInterrupted()) {
            moduleThread.interrupt();
            moduleThread = null;
        }
        moduleDel(temp);
    }

}

File: sdk/binding/android/q-voice/src/main/java/com/qdreamer/qvoice/QModule.kt
package com.qdreamer.qvoice

import com.qdreamer.qvoice.listener.OnQVoiceCallback
import kotlinx.coroutines.*

class QModule(
    sessionValue: Long, param: String, resPath: String, private val callback: OnQVoiceCallback?
) {

    private companion object {
        private const val MODULE_VALUE = 0L
    }

    // private native long moduleNew(long sessionValue, String param, String resPath);
    private external fun moduleNew(sessionValue: Long, param: String, resPath: String): Long
    private external fun moduleDel(moduleValue: Long)
    private external fun moduleStart(moduleValue: Long): Int
    private external fun moduleSet(moduleValue: Long, params: String): Int
    private external fun moduleStop(moduleValue: Long): Int
    private external fun moduleRead(moduleValue: Long): ByteArray?
    private external fun moduleFeed(moduleValue: Long, data: ByteArray, len: Int = data.size): Int

    private var moduleValue: Long = MODULE_VALUE
    private var moduleJob: Job? = null

    private var isModuleStarted = false

    init {
        moduleValue = moduleNew(sessionValue, param, resPath)
        moduleJob?.cancel()
        moduleJob = GlobalScope.launch(Dispatchers.IO) {
            while (isActive && moduleValue != MODULE_VALUE) {
                val bytes = moduleRead(moduleValue)
                if (isActive && bytes != null && bytes.isNotEmpty()) {
                    if (bytes.size > 1) {
                        callback?.onQVoiceCallback(bytes[0], bytes.copyOfRange(1, bytes.size))
                    } else {
                        callback?.onQVoiceCallback(bytes[0], null)
                    }
                }
                Thread.sleep(1)
            }
        }
    }

    fun startModule(): Boolean {
        synchronized(isModuleStarted) {
            if (moduleValue != MODULE_VALUE && !isModuleStarted) {
                moduleStart(moduleValue)
                isModuleStarted = true
            }
            return isModuleStarted
        }
    }

    fun feedModule(audio: ByteArray) {
        if (moduleValue != MODULE_VALUE && isModuleStarted) {
            moduleFeed(moduleValue, audio, audio.size)
        }
    }

    fun stopModule(): Boolean {
        synchronized(isModuleStarted) {
            if (moduleValue != MODULE_VALUE && isModuleStarted) {
                moduleStop(moduleValue)
                isModuleStarted = false
            }
            return !isModuleStarted
        }
    }

    fun releaseModule() {
        stopModule()
        moduleJob?.cancel()
        val temp = moduleValue
        moduleValue = MODULE_VALUE
        moduleDel(temp)
    }

}